rest/route/pod_agent.go

```diff
@@ -328,3 +348,155 @@ func (h *podAgentNextTask) findDispatcher() (*dispatcher.PodDispatcher, error) {
+
+	if p.RunningTask == "" {
+		grip.Notice(message.Fields{
+			"message":                 "pod is not assigned task, not clearing, asking agent to exit",
+			"task_id":                 t.Id,
+			"task_status_from_db":     t.Status,
+			"task_details_from_db":    t.Details,
+			"current_agent":           p.AgentVersion == evergreen.AgentVersion,
+			"agent_version":           p.AgentVersion,
+			"build_revision":          evergreen.BuildRevision,
+			"build_agent":             evergreen.AgentVersion,
+			"task_details_from_agent": h.details,
+			"pod_id":                  p.ID,
+		})
+		endTaskResp.ShouldExit = true
+		return gimlet.NewJSONResponse(endTaskResp)
+	}
+
+	// Clear the running task on the pod now that the task has finished.
+	if err := p.ClearRunningTask(); err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "clearing running task %s for pod %s", t.Id, p.ID))
+	}
+
+	projectRef, err := model.FindMergedProjectRef(t.Project, t.Version, true)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusInternalServerError,
+			Message:    errors.Wrap(err, "finding project").Error(),
+		})
+	}
+	if projectRef == nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusNotFound,
+			Message:    errors.Errorf("project '%s' not found", t.Project).Error(),
+		})
+	}
+
+	// mark task as finished
+	deactivatePrevious := utility.FromBoolPtr(projectRef.DeactivatePrevious)
+	err = model.MarkEnd(t, evergreen.APIServerTaskActivator, finishTime, &h.details, deactivatePrevious)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "calling mark finish on task %v", t.Id))
+	}
+
+	if t.Requester == evergreen.MergeTestRequester {
+		if err = model.HandleEndTaskForCommitQueueTask(t, h.details.Status); err != nil {
+			return gimlet.MakeJSONErrorResponder(err)
+		}
+	}
+
+	// the task was aborted if it is still in undispatched.
+	// the active state should be inactive.
+	if h.details.Status == evergreen.TaskUndispatched {
+		if t.Activated {
+			grip.Warningf("task %v is active and undispatched after being marked as finished", t.Id)
+			return gimlet.NewJSONResponse(&apimodels.EndTaskResponse{})
+		}
+		grip.Infof(fmt.Sprintf("task %v has been aborted and will not run", t.Id))
+		return gimlet.NewJSONResponse(&apimodels.EndTaskResponse{})
+	}
+
+	// GetDisplayTask will set the DisplayTask on t if applicable
+	// we set this before the collect task end job is run to prevent data race
+	dt, err := t.GetDisplayTask()
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "getting display task"))
+	}
+	queue := h.env.RemoteQueue()
+	job := units.NewCollectTaskEndDataJob(t, nil, p)
+	if err = queue.Put(ctx, job); err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "couldn't queue job to update task stats accounting"))
+	}
+
+	if p.Status != pod.StatusRunning {
+		endTaskResp.ShouldExit = true
```

Comment:
Same thing as above - if it's not running, the move should be to enqueue the termination job to stop the pod.